pkg/blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,542 @@
+			"num_retries", boff.NumRetries(),
+		)
+		boff.Wait()
+	}
+	return boff.Err()
+}
+
+func (b *BlockBuilder) stopping(_ error) error {
+	b.kafkaClient.Close()
+
+	return nil
+}
+
+func (b *BlockBuilder) running(ctx context.Context) error {
+	// Do initial consumption on start using current time as the point up to which we are consuming.
+	// To avoid small blocks at startup, we consume until the last hour boundary + buffer.
+	cycleEnd := cycleEndAtStartup(b.cfg.ConsumeInterval, b.cfg.ConsumeIntervalBuffer)
+	err := b.NextConsumeCycle(ctx, cycleEnd)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return err
+	}
+
+	nextCycleTime := time.Now().Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+	waitTime := time.Until(nextCycleTime)
+	for waitTime > b.cfg.ConsumeInterval {
+		// NB: at now=14:12, next cycle starts at 14:15 (startup cycle ended at 13:15)
+		//     at now=14:17, next cycle starts at 15:15 (startup cycle ended at 14:15)
+		nextCycleTime = nextCycleTime.Add(-b.cfg.ConsumeInterval)
+		waitTime -= b.cfg.ConsumeInterval
+	}
+
+	for {
+		select {
+		case <-time.After(waitTime):
+			cycleEnd := nextCycleTime
+			level.Info(b.logger).Log("msg", "triggering next consume from running", "cycle_end", cycleEnd, "cycle_time", nextCycleTime)
+			err := b.NextConsumeCycle(ctx, cycleEnd)
+			if err != nil {
+				b.metrics.consumeCycleFailures.Inc()
+				level.Error(b.logger).Log("msg", "consume cycle failed", "cycle_end", cycleEnd, "err", err)
+			}
+
+			// If we took more than consumptionItvl time to consume the records, this will immediately start the next consumption.
+			// TODO(codesome): track waitTime < 0, which is the time we ran over. Should have an alert on this.
+			nextCycleTime = nextCycleTime.Add(b.cfg.ConsumeInterval)
+			waitTime = time.Until(nextCycleTime)
+		case <-ctx.Done():
+			level.Info(b.logger).Log("msg", "context cancelled, stopping block builder")
+			return nil
+		}
+	}
+}
+
+func cycleEndAtStartup(interval, buffer time.Duration) time.Time {
+	cycleEnd := time.Now().Truncate(interval).Add(buffer)
+	if cycleEnd.After(time.Now()) {
+		cycleEnd = cycleEnd.Add(-interval)
+	}
+	return cycleEnd
+}
+
+// NextConsumeCycle manages consumption of currently assigned partitions.
+// The cycleEnd argument indicates the timestamp (relative to Kafka records) up until which to consume from partitions
+// in this cycle. That is, Kafka records produced after the cycleEnd mark will be consumed in the next cycle.
+func (b *BlockBuilder) NextConsumeCycle(ctx context.Context, cycleEnd time.Time) error {
+	defer func(t time.Time) {
+		b.metrics.consumeCycleDuration.Observe(time.Since(t).Seconds())
+	}(time.Now())
+
+	for _, part := range b.parts {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if part < 0 {
+			return nil
+		}
+
+		// TODO(v): calculating lag for each individual partition requests data for the whole group every time. This is redundant.
```

Comment:
i agree we do a lot of work multiple times and throw away most of it. But wouldn't lag change as we move along the partitions? For example the lag now might be lower than the lag after we process one partition.

Also - this should happen fairly infrequently, so doing the work multiple times and throwing away the result may be ok.

Reply:
> wouldn't lag change as we move along the partitions? For example the lag now might be lower than the lag after we process one partition

Yes, but also note that block-builder's cycles consume data from the past hour, up until beginning of this hour. I.e. if cycle starts at time `T`, block-builder needs to consume records produced in interval `[T-1h, T)`. For us, records produced after `T` (i.e. during the cycle) belong to the next cycle.

There is an edge case, when someone produces Kafka records with time set in the past. We cannot gracefully handle that in the current iteration, so those will be consumed in the next cycle anyway.

Reply:
aha, that makes sense. So caching the lag for each consume cycle can only potentially delay ingesting some data until the next cycle. But that's probably not a big deal. Yeah I think what you have is easier to reason about.